### ISSUE TITLE
Image not loading after upgrading to 4.1.0

### ISSUE THREAD

**Issue by airam85** [openseadragon/openseadragon#2371]
I have a Vue2 project using openseadragon 2.4.2 (installed with npm) and it was working correctly. 

I've tried upgrading to 4.1.0 and the image doesn't load, only the navigator minimap. If I don't show the navigator, the image loads correctly, do yo have any idea why? I've also discover it stopped working in 4.0.0, in previous versions the image loads just fine.
If I try to do the same with vanilla javascript it is working correctly, I don't know if it is something in Vue interfering with Openseadragon or what could this be.

**Comment by iangilman** [comment]
My guess is your issue has to do with https://github.com/openseadragon/openseadragon/pull/2256 by @pearcetm. Is there a moment at the beginning where the viewer is extremely small or oddly shaped? If so, when it then gets resized to its proper size, it doesn't auto-resize correctly. This is a side effect of what was otherwise a good improvement. If you could, it would be great to have a repro (codepen or such). 

Meanwhile, I think the workaround is to either not have that moment where the viewer is badly sized, or do a `viewer.viewport.goHome(true)` after the viewer is sized correctly.

**Comment by airam85** [comment]
I initialize the viewer in the mounted state of the app, I thought that meant the div is already its size, but maybe it is not yet? 
I've made a really simple repo for you to see the issue: https://github.com/airam85/openseadragon-vue-example
Couldn't make it work in codepen, I don't know why there it wasn't failling.

**Comment by pearcetm** [comment]
> Couldn't make it work in codepen, I don't know why it wasn't failling there.

Does this mean that your code when run in codepen opens the image correctly in the viewer and the navigator? If so, what is different about whatever environment you are using to run it?

**Comment by airam85** [comment]
> Does this mean that your code when run in codepen opens the image correctly in the viewer and the navigator? If so, what is different about whatever environment you are using to run it?

Yes, I've tried to replicate the code here,  but since I'm using Vue the code it is not exactly the same: https://codepen.io/airam1985/pen/xxQbOKm
In the github repository, the code loading OSD is inside an App.vue file, here is all in the js part of the code, and in the repository I run it with the command: npm run serve. I don't know if any of these differences are what  (I can't think of anything more that is relevant).

**Comment by pearcetm** [comment]
I see - yes, it definitely seems to be working in the codepen.

I wonder if this note from the vue docs might have something to do with the behavior you're seeing: https://v2.vuejs.org/v2/api/#mounted

> mounted: `function () {  this.$nextTick(function () {    // Code that will run only after the    // entire view has been rendered  })}`

> Note that mounted does not guarantee that all child components have also been mounted. If you want to wait until the entire view has been rendered, you can use [vm.$nextTick](https://v2.vuejs.org/v2/api/#vm-nextTick) inside of mounted:
```
mounted: function () {
  this.$nextTick(function () {
    // Code that will run only after the
    // entire view has been rendered
  })
}
```

I tried cloning your repo but `npm install` is giving a bunch of errors so I can't test locally

**Comment by airam85** [comment]
> I wonder if this note from the vue docs might have something to do with the behavior you're seeing: https://v2.vuejs.org/v2/api/#mounted
Yeah, I've seen that too, and tried to put it inside the nextTick,  but the error persisted. Also, in my original project, the OSD is initlialized after loading some data from the server, so it is not just when the component is mounted but a few seconds after.

> I tried cloning your repo but `npm install` is giving a bunch of errors so I can't test locally
I think it is because I have an old node version, I've uploaded a Dockerfile if you want to try to deploy it that way. If I have more time another day I'll try to update my node version and modify the project so it runs.

**Comment by iangilman** [comment]
@pearcetm Wow! That's a subtle one. Thank you for tracking it down! 

Would it make sense to replace our `$.extend` entirely with [Object.assign()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)? I assume we didn't use it originally for IE compatibility reasons, but that's not an issue anymore.

**Comment by pearcetm** [comment]
> Wow! That's a subtle one. Thank you for tracking it down!

Yeah, that took some digging to find...



> Would it make sense to replace our `$.extend` entirely with [Object.assign()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)? I assume we didn't use it originally for IE compatibility reasons, but that's not an issue anymore.

I don't think this makes sense, at least not in a naive implementation - Object.assign won't make a deep copy, and will copy objects (including things like arrays) by reference, which is not what we want. The library would break in innumerable ways.

The problem leading to the current issue/question (and the related question https://github.com/openseadragon/openseadragon/issues/2234#issuecomment-1588569744) is that the Vue.js framework (v2) converts all fields of objects to have getters/setters, which changes the behavior of OpenSeadragon. From my cursory reading it seems that Vue v3 uses a different technique, and might be more compatible with OSD (but I haven't tested this, I don't use Vue for anything). Perhaps in the application code the Vue framework could be instructed not to touch the OpenSeadragon object? Maybe there's a compelling argument to modify OSD here, but I'd favor putting the burden of compatibility on the library that is unintentionally changing the behavior of OSD, rather than trying to build features into OSD to make it robust to other frameworks trying to modify it.

**Comment by pearcetm** [comment]
@airam85 @wdj577 Would either of you be able to test if one of the following approaches work for you?

1) Disable Vue tracking the OpenSeadragon object.
2) Upgrade to Vue 3.0+ (uses `Proxy` instead of `Object.defineProperty` to watch and mutate objects)

**Comment by wdj577** [comment]
> 

@pearcetm I used the first method of adding Vue.config.ignoredElements = ['openseadragon'] in the main.js file of my Vue project, but it still didn't work. Then, I tried the second method by creating a new Vue 3 project and importing version 4.1.0 for previewing the "zoomifytileservice" type of images, and it worked without any issues! This strongly suggests that some mechanisms in Vue 2 are affecting it.
vue3demo:
[test.zip](https://github.com/openseadragon/openseadragon/files/11767622/test.zip)

**Comment by wdj577** [comment]
@airam85 
I tried integrating your demo DZI images into my project, and indeed, they worked fine! This proves that different build tools can also have some influence.

![image](https://github.com/openseadragon/openseadragon/assets/22167740/6dacd810-b030-4519-8bd1-8aaea2fe0e14)
I may not have enough understanding of the underlying workings of Vue.How could this happen?

**Comment by iangilman** [comment]
@pearcetm Right, I hadn't thought about the deep copy! I think `Object.assign` would be a good choice in places where we don't need deep copy though. It looks like there are 53 uses of `$.extend` in total and just 16 uses of `$.extend( true`. I agree we shouldn't be going overboard to make specific changes for certain frameworks, but in this case that seems like nice cleanup anyway. 

Where is the `extend` that's causing trouble? Is it this one? 

https://github.com/openseadragon/openseadragon/blob/ebab356c207e626b6622f88ffcb0cd28b918f85d/src/navigator.js#L428

Or this? 

https://github.com/openseadragon/openseadragon/blob/ebab356c207e626b6622f88ffcb0cd28b918f85d/src/viewer.js#L1626

Neither of those are deep copies. 

BTW, if we do come up with a good recommendation for Vue, sounds like another addition for the FAQ!

**Comment by pearcetm** [comment]
> Where is the `extend` that's causing trouble? Is it this one?

It's this one: https://github.com/openseadragon/openseadragon/blob/ebab356c207e626b6622f88ffcb0cd28b918f85d/src/viewer.js#L1626

Replacing `$.extend` with `Object.assign` here does fix the issue in the example code I have using Vue2.


> I think `Object.assign` would be a good choice in places where we don't need deep copy though.

It works in this particular case, but it doesn't copy getters/setters so it may not always behave intuitively... I guess I'm not sure how to know whether it would solve more problems than it could potentially cause.


> BTW, if we do come up with a good recommendation for Vue, sounds like another addition for the FAQ!

Agreed! I still favor figuring out how to get Vue to leave core OSD alone (perhaps with the ability to watch/bind to certain properties only).

**Comment by wdj577** [comment]
> Replacing `$.extend` with `Object.assign` here does fix the issue in the example code I have using Vue2.

Absolutely brilliant! The problem has been effectively resolved after implementing this solution. I must express my deepest gratitude for your invaluable technical support. Thank you ever so much!

**Comment by pearcetm** [comment]
> I've tried to create a new project with vue3 and now It doesn't load anything, not even the navigator.

@airam85 You've been bitten by the fact that OSD needs to be given defined size (https://github.com/openseadragon/openseadragon/issues/2368#issuecomment-1579463448). See https://github.com/airam85/osd-vue3/issues/1 for a fix in your CSS. With this fixed, Vue3 seems to work just fine with OSD.

@iangilman With Vue3 working fine with OSD 4.0.0+ I would vote that we add this compatibility info to the upcoming FAQ (i.e. Vue2 is not compatible with 4.0.0+ (and we could give the hint about `Object.assign`), and Vue3+ should be used instead)

**Comment by airam85** [comment]
> @airam85 You've been bitten by the fact that OSD needs to be given defined size ([#2368 (comment)](https://github.com/openseadragon/openseadragon/issues/2368#issuecomment-1579463448)). See [airam85/osd-vue3#1](https://github.com/airam85/osd-vue3/issues/1) for a fix in your CSS. With this fixed, Vue3 seems to work just fine with OSD.

Oh, thanks! I'm going to migrate to vue3, then!

**Comment by iangilman** [comment]
> It works in this particular case, but it doesn't copy getters/setters so it may not always behave intuitively... I guess I'm not sure how to know whether it would solve more problems than it could potentially cause.

Good point!

> With Vue3 working fine with OSD 4.0.0+ I would vote that we add this compatibility info to the upcoming FAQ (i.e. Vue2 is not compatible with 4.0.0+ (and we could give the hint about Object.assign), and Vue3+ should be used instead)

Sounds good! How about we get the FAQ started on the wiki here?

https://github.com/openseadragon/openseadragon/wiki/FAQ

Maybe we move it to the main site eventually, or maybe the wiki is a good place for it so it can be a bit more dynamic. Either way, we want to link to it from various places on the main site.
